Provide module-info.java for Java9+ users

I know you already provide an Automatic-Module-Name for Java9+ users but we would be better off with a multi-release JAR that exports a full-fledged `module-info.java` because when using `requires transitive org.jsoup` the compiler complains (warns) against exporting automatic modules.

Publishing a MultiRelease JAR using Maven should be relatively easy. See `multiReleaseOutput` at https://www.baeldung.com/maven-multi-release-jars

Hit me up if you run into any problems and I'll take a look.

Hi @cowwoc  - I gave this a try, but can't get it to work. Animal Sniffer is failing when it tests (either for Java 1.8 or for Android 10 signature -- the exception doesn't specify which signature is failing on which class.

Have pushed a PR - #1470. If you have any ideas I'd love to hear them. I flailed around in Maven trying to upgrade dependencies and trying to find similar issues but could not. I can't find any projects that have a multi release jar and are using Animal Sniffer, if we could find a reference I think that would probably solve the issue.

```text
[ERROR] Failed to execute goal org.codehaus.mojo:animal-sniffer-maven-plugin:1.19:check (compile-java-8) on project jsoup: Execution compile-java-8 of goal org.codehaus.mojo:animal-sniffer-maven-plugin:1.19:check failed: This feature
requires ASM6 -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:animal-sniffer-maven-plugin:1.19:check (compile-java-8) on project jsoup: Execution compile-java-8 of goal org.codehaus.mojo:animal-sniff
er-maven-plugin:1.19:check failed: This feature requires ASM6
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:567)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution compile-java-8 of goal org.codehaus.mojo:animal-sniffer-maven-plugin:1.19:check failed: This feature requires ASM6
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:567)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.UnsupportedOperationException: This feature requires ASM6
    at org.objectweb.asm.ClassVisitor.visitModule (ClassVisitor.java:137)
    at org.objectweb.asm.ClassReader.readModuleAttributes (ClassReader.java:754)
    at org.objectweb.asm.ClassReader.accept (ClassReader.java:553)
    at org.objectweb.asm.ClassReader.accept (ClassReader.java:401)
    at org.codehaus.mojo.animal_sniffer.ClassListBuilder.process (ClassListBuilder.java:70)
    at org.codehaus.mojo.animal_sniffer.ClassFileVisitor.processClassFile (ClassFileVisitor.java:197)
    at org.codehaus.mojo.animal_sniffer.ClassFileVisitor.process (ClassFileVisitor.java:116)
    at org.codehaus.mojo.animal_sniffer.ClassFileVisitor.process (ClassFileVisitor.java:96)
    at org.codehaus.mojo.animal_sniffer.ClassFileVisitor.processDirectory (ClassFileVisitor.java:134)
    at org.codehaus.mojo.animal_sniffer.ClassFileVisitor.process (ClassFileVisitor.java:112)
    at org.codehaus.mojo.animal_sniffer.ClassFileVisitor.process (ClassFileVisitor.java:96)
    at org.codehaus.mojo.animal_sniffer.ClassFileVisitor.processDirectory (ClassFileVisitor.java:134)
    at org.codehaus.mojo.animal_sniffer.ClassFileVisitor.process (ClassFileVisitor.java:112)
    at org.codehaus.mojo.animal_sniffer.ClassFileVisitor.process (ClassFileVisitor.java:96)
    at org.codehaus.mojo.animal_sniffer.ClassFileVisitor.processDirectory (ClassFileVisitor.java:134)
    at org.codehaus.mojo.animal_sniffer.ClassFileVisitor.process (ClassFileVisitor.java:112)
    at org.codehaus.mojo.animal_sniffer.ClassFileVisitor.process (ClassFileVisitor.java:96)
    at org.codehaus.mojo.animal_sniffer.ClassFileVisitor.processDirectory (ClassFileVisitor.java:134)
    at org.codehaus.mojo.animal_sniffer.ClassFileVisitor.process (ClassFileVisitor.java:112)
    at org.codehaus.mojo.animal_sniffer.maven.CheckSignatureMojo.apply (CheckSignatureMojo.java:366)
    at org.codehaus.mojo.animal_sniffer.maven.CheckSignatureMojo.buildPackageList (CheckSignatureMojo.java:359)
    at org.codehaus.mojo.animal_sniffer.maven.CheckSignatureMojo.execute (CheckSignatureMojo.java:251)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:567)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[ERROR]
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException

```

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -12,7 +12,7 @@ jobs:
       matrix:
         os: [ubuntu-latest, windows-latest, macOS-latest]
         # choosing to run a reduced set of LTS, current, and next, to balance coverage and execution time
-        java: [9, 17, 21]
+        java: [11, 17, 21]
       fail-fast: false
     name: Test JDK ${{ matrix.java }}, ${{ matrix.os }}
     steps:

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -12,7 +12,7 @@ jobs:
       matrix:
         os: [ubuntu-latest, windows-latest, macOS-latest]
         # choosing to run a reduced set of LTS, current, and next, to balance coverage and execution time
-        java: [8, 17, 21]
+        java: [9, 17, 21]
       fail-fast: false
     name: Test JDK ${{ matrix.java }}, ${{ matrix.os }}
     steps:

--- a/CHANGES
+++ b/CHANGES
@@ -11,6 +11,8 @@ Release 1.17.1 [PENDING]
   * Improvement: added the `:is(selector list)` pseudo-selector, which finds elements that match any of the selectors in
     the selector list. Useful for making large ORed selectors more readable.
 
+  * Improvement: repackaged the library with native (vs automatic) JPMS module support.
+
   * Bugfix: when outputting with XML syntax, HTML elements that were parsed as data nodes (<script> and <style>) should
     be emitted as CDATA nodes, so that they can be parsed correctly by an XML parser.
     <https://github.com/jhy/jsoup/pull/1720>

--- a/CHANGES
+++ b/CHANGES
@@ -12,6 +12,7 @@ Release 1.17.1 [PENDING]
     the selector list. Useful for making large ORed selectors more readable.
 
   * Improvement: repackaged the library with native (vs automatic) JPMS module support.
+    <https://github.com/jhy/jsoup/pull/2025>
 
   * Bugfix: when outputting with XML syntax, HTML elements that were parsed as data nodes (<script> and <style>) should
     be emitted as CDATA nodes, so that they can be parsed correctly by an XML parser.

--- a/pom.xml
+++ b/pom.xml
@@ -38,13 +38,12 @@
 
   <build>
     <plugins>
+      <!-- Compile -->
       <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-compiler-plugin</artifactId>
         <version>3.11.0</version>
         <configuration>
-          <source>1.8</source>
-          <target>1.8</target>
           <encoding>UTF-8</encoding>
           <compilerArgs>
             <!-- saves output for package-info.java, so mvn sees it has completed it, so incremental compile works -->
@@ -53,9 +52,34 @@
           <!-- this means incremental = true... -->
           <useIncrementalCompilation>false</useIncrementalCompilation>
         </configuration>
+        <executions>
+          <execution>
+            <id>compile-java-8</id>
+            <configuration>
+              <source>1.8</source>
+              <target>1.8</target>
+            </configuration>
+          </execution>
+          <execution>
+            <id>compile-java-9</id>
+            <phase>compile</phase>
+            <goals>
+              <goal>compile</goal>
+            </goals>
+            <configuration>
+              <release>9</release>
+              <compileSourceRoots>
+                <compileSourceRoot>${project.basedir}/src/main/java9</compileSourceRoot>
+              </compileSourceRoots>
+              <multiReleaseOutput>true</multiReleaseOutput>
+            </configuration>
+          </execution>
+        </executions>
+
       </plugin>
+
+      <!-- Ensure Java 8 and Android 10 API compatibility -->
       <plugin>
-        <!-- Ensure Java 8 and Android 10 API compatibility -->
         <groupId>org.codehaus.mojo</groupId>
         <artifactId>animal-sniffer-maven-plugin</artifactId>
         <version>1.23</version>
@@ -136,7 +160,7 @@
               <addDefaultImplementationEntries>true</addDefaultImplementationEntries>
             </manifest>
             <manifestEntries>
-              <Automatic-Module-Name>org.jsoup</Automatic-Module-Name>
+              <Multi-Release>true</Multi-Release>
             </manifestEntries>
             <manifestFile>${project.build.outputDirectory}/META-INF/MANIFEST.MF</manifestFile>
           </archive>

--- a/src/main/java9/module-info.java
+++ b/src/main/java9/module-info.java
@@ -1,0 +1,10 @@
+module org.jsoup {
+    exports org.jsoup;
+    exports org.jsoup.helper;
+    exports org.jsoup.nodes;
+    exports org.jsoup.parser;
+    exports org.jsoup.safety;
+    exports org.jsoup.select;
+
+    requires transitive java.xml; // for org.w3c.dom out of W3CDom
+}

